libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -383,11 +383,10 @@ string IRGenerator::memoryInit()
 void IRGenerator::resetContext(ContractDefinition const& _contract)
 {
 	solAssert(
-		m_context.functionCollector()->requestedFunctions().empty(),
+		m_context.functionCollector().requestedFunctions().empty(),
 		"Reset context while it still had functions."
 	);
 	m_context = IRGenerationContext(m_evmVersion, m_context.revertStrings(), m_optimiserSettings);
-	m_utils = YulUtilFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
```

Comment:
There's actually no reason for resetting this here. If *not* resetting feels too dangerous, though, we have several options:
We can make the constructor ``noexcept`` and use:
```
m_utils.~YulUtilFunctions();
new (&m_utils) YulUtilFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
```
Or if that is too exotic, we can store it in a ``unique_ptr`` here and recreate that or alternatively add a ``reset`` function.

Resetting as we did before won't work, since there can be no assignment operator, if we store the function collector as reference.

Reply:
instead of manually calling those two lines (and because people will ponder each time they see this trick), it would make sense to have such a functionality as a generic helper function in our libsolutil, isn't it?

```cpp
template <typename T, typename... Args>
void recreate(T& obj, Args&&... args)
{
    obj.~T();
    new (&obj) T(std::foward<T>(args)...);
}
```
What do you think?

Reply:
It's only safe if the constructor is noexcept and actually I'd hardly ever make a constructor noexcept, even if it *was* possible - I don't think we should go that way... the main question is whether we just *don't* reset, because there in fact is nothing to reset anymore after this change.

Reply:
Actually, I'm not even sure, the destructor-then-placement-new-thing is even strictly speaking allowed in the presence of a member reference :-).